python-lib/ocr_utils.py

```diff
@@ -80,3 +87,47 @@ def get_default_ocr_engine():
     if which("tesseract") is not None:  # check if tesseract is in the path
         return Constants.TESSERACT
     return Constants.EASYOCR
+
+
+def try_download_pandoc():
+    """ download pandoc prebuilt binaries into the temporary job folder """
+    try:
+        import pypandoc
+        pandoc_tmp_directory = os.getcwd()
+        pypandoc.download_pandoc(targetfolder=pandoc_tmp_directory, download_folder=pandoc_tmp_directory)
+        return True
+    except Exception as e:
+        logger.warning("Failed to download pandoc binaries: {}".format(e))
+        return False
+
+
+def extract_text(file_bytes, extension, with_pandoc):
+    """
+    First extract with pdfium or docx PDFs and docx files.
+    Then try to extract text using pandoc into plain text.
+    Finally, just decode the bytes if pandoc failed or is not downloaded.
+    """
+    if extension == "pdf":
+        pdf_pages = pdfium.PdfDocument(file_bytes)
+        return "\n".join([page.get_textpage().get_text_range() for page in pdf_pages])
+    elif extension == "docx":
+        doc = docx.Document(BytesIO(file_bytes))
+        return "\n".join([paragraph.text for paragraph in doc.paragraphs])
+    else:
+        text = ""
+        if with_pandoc:
+            try:
+                temporary_job_folder = os.getcwd()
+                with tempfile.NamedTemporaryFile(dir=temporary_job_folder, suffix=".{}".format(extension)) as tmp:
+                    tmp.write(file_bytes)
+                    text = pypandoc.convert_file(tmp.name, to="plain", format=extension)
```

Comment:
Does this work if extension is `""` ? or `None` ?

Reply:
`pandoc` has a method `_validate_formats` to check the format and it will raise custom errors for bad or unsupported formats.